Sources/XMLCoder/Box/DictionaryBox.swift

```diff
@@ -0,0 +1,105 @@
+//
+//  DictionaryBox.swift
+//  XMLCoderPackageDescription
+//
+//  Created by Vincent Esche on 11/19/18.
+//
+
+import Foundation
+
+// Minimalist implementation of an order-preserving keyed box:
+internal class DictionaryBox {
+    typealias Key = String
+    typealias Value = Box
+    
+    typealias Unboxed = [Key: Value]
+    
+    private(set) var unboxed: Unboxed
+    
+    var count: Int {
+        return self.unboxed.count
+    }
+    
+    var keys: Unboxed.Keys {
+        return self.unboxed.keys
+    }
+    
+    subscript(key: Key) -> Value? {
+        get {
+            return self.unboxed[key]
+        }
+        set {
+            self.unboxed[key] = newValue
+        }
+    }
+    
+    init(_ unboxed: Unboxed = [:]) {
+        self.unboxed = unboxed
+    }
+    
+    convenience init<S>(_ keysAndValues: S, uniquingKeysWith combine: (Value, Value) throws -> Value) rethrows where S : Sequence, S.Element == (Key, Value) {
+        let unboxed = try Dictionary(keysAndValues, uniquingKeysWith: combine)
+        self.init(unboxed)
+    }
+    
+    func unbox() -> Unboxed {
+        return self.unboxed
+    }
+    
+    func filter(_ isIncluded: (Key, Value) throws -> Bool) rethrows -> [(Key, Value)] {
+        return try self.unboxed.filter(isIncluded)
+    }
+    
+    func map<T>(_ transform: (Key, Value) throws -> T) rethrows -> [T] {
+        return try self.unboxed.map(transform)
+    }
+    
+    func compactMap<T>(_ transform: ((Key, Value)) throws -> T?) rethrows -> [T] {
+        return try self.unboxed.compactMap(transform)
+    }
+    
+    func mapValues(_ transform: (Value) throws -> Value) rethrows -> DictionaryBox {
+        return DictionaryBox(try self.unboxed.mapValues(transform))
+    }
+}
+
+extension DictionaryBox: Box {
+    var isNull: Bool {
+        return false
+    }
+    
+    var isFragment: Bool {
+        return false
+    }
+    
+    var xmlString: String? {
+        return nil
+        
+//        let elements = self.unboxed.sorted { $0.0 < $1.0 }
+//        let strings: [String] = elements.compactMap { key, box in
+//            switch box {
+//            case .string(let box):
+//                return "'\(key)': '\(box.xmlString)'"
+//            case _:
+//                if let value = box.xmlString {
+//                    return "'\(key)': \(value)"
+//                } else {
+//                    return nil
+//                }
+//            }
+//        }
+//        return "[" + strings.joined(separator: ", ") + "]"
+    }
+}
+
+//extension DictionaryBox: Equatable {
```

Comment:
commented extension? Which can probably be derived automatically by the compiler if needed

Reply:
It can't any more, due to `Box` not being `Box: Equatable`. The latter would turn it into a PAT, rendering it unusable for our purposes.

But I only needed the `Equatable` for testing anyway. Using a workaround there now.

Reply:
that makes sense, thank you. Would be great if the workaround in the test has a comment explaining that it exists only because of the PAT issue so that in the future we don't forget why it was done that way.

Reply:
I'm basically doing this:

```swift
func testUnbox() {
    let unboxed = box.unbox()
    XCTAssertEqual(unboxed.count, 2)
    XCTAssertEqual(unboxed[0] as? StringBox, StringBox("foo"))
    XCTAssertEqual(unboxed[1] as? IntBox, IntBox(42))
}
```

rather than this:

```swift
func testUnbox() {
    let unboxed = box.unbox()
    XCTAssertEqual(unboxed, [StringBox("foo"), IntBox(42)])
}
```

No magic in there. Just avoiding the type-erasure of `Box`.